File: nuuwatch/src/modules/anime/jikan/jikan_settings.rs
use url::Url;

pub struct Settings {
    base_url: Url,
}

impl Settings {
    pub fn new(url: &str) -> Self {
        let mut url = Url::parse(url).unwrap();
        // Default to v4 if not set in constructor, or pass version arg
        Settings { base_url: url}
    }

    pub fn get_base_url(&self) -> Url {
        self.base_url.clone()
    }
}

impl Default for Settings {
    fn default() -> Self {
        Settings::new("https://api.jikan.moe/v4")
    }
}

File: nuuwatch/frontend/js/jikan/jikan-settings.ts
export class Settings {
  protocol = 'https:';
  private baseURL!: URL;
  private v: number | undefined;

  constructor(baseURL = 'https://api.jikan.moe', version = 4) {
    this.setBaseURL(baseURL, version);
  }

  /**
   * Delivers the full API Base URL
   * @returns {URL}
   */
  getBaseURL() {
    return this.baseURL;
  }

  /**
   * can be used to replace the current API Base URL by a complete new one
   * @param {string} baseURL
   * @param {number} [version]
   */
  setBaseURL(baseURL: string, version?: number) {
    if(version) this.v = version;
    this.baseURL = new URL(`/v${this.v}`, baseURL);
  }

  setBaseURLAbsolute(pathname: string, origin: string = location.origin) {
    this.baseURL = new URL(pathname, origin);

    this.protocol = this.baseURL.protocol;
  }

  /**
   * can be used to change the API version
   * @param {number} version
   */
  set version(version) {
    this.v = version;
    this.baseURL.pathname = `/v${version}`;
  }

  /**
   * delivers the currently used API version
   * @returns {number}
   */
  get version() {
    return this.v;
  }
}

export const settings = new Settings();
